Sources/_MatchingEngine/Regex/Parse/Mocking.swift

```diff
@@ -1,107 +1,142 @@
+    switch self {
+    case .regular: return ("'/", "/'")
+    case .modern: return ("'|", "|'")
     }
-    fatalError()
   }
-
-  static func lexSingleQuotedLiteral(_ sub: Substring) -> Tok {
-    let ast: AST
-    if sub.first == "/" && sub.last == "/" {
-      ast = try! parse(
-        sub.dropFirst().dropLast(), .traditional)
-    } else {
-      ast = try! parse(sub, .modern)
+  public var opening: String { openingAndClosing.opening }
+  public var closing: String { openingAndClosing.closing }
+
+  /// The default set of syntax options that the delimiter indicates.
+  public var defaultSyntaxOptions: SyntaxOptions {
+    switch self {
+    case .regular: return .traditional
+    case .modern: return .modern
     }
+  }
+}
 
-    // TODO: ... call regex parser
-    return .opaque(ast)
+public struct LexError: Error, CustomStringConvertible {
+  enum Kind: Hashable {
+    case endOfString
+    case invalidUTF8 // TODO: better range reporting
+    case unknownDelimiter
   }
 
-}
+  var kind: Kind
 
-// TODO: mock up multi-line soon
+  /// The pointer at which to resume lexing.
+  public var resumePtr: UnsafeRawPointer
 
+  init(_ kind: Kind, resumeAt resumePtr: UnsafeRawPointer) {
+    self.kind = kind
+    self.resumePtr = resumePtr
+  }
 
-enum LexError: Error {
-  case endOfString
-  case invalidUTF8 // TODO: better range reporting
-  case notRegexLiteralStart
+  /// Whether the lexer came across an unknown delimiter, indicating that this
+  /// is not in fact a regex literal.
+  public var isUnknownDelimiter: Bool { kind == .unknownDelimiter }
+
+  public var description: String {
+    switch kind {
+    case .endOfString: return "unterminated regex literal"
+    case .invalidUTF8: return "invalid UTF-8 found in source file"
+    case .unknownDelimiter: return "unknown regex literal delimiter"
+    }
+  }
 }
 
+/// Drop a set of regex delimiters from the input string, returning the contents
+/// and the delimiter used. The input string must have valid delimiters.
+func droppingRegexDelimiters(_ str: String) -> (String, Delimiter) {
+  let utf8 = str.utf8
+  func stripDelimiter(_ delim: Delimiter) -> String? {
+    let prefix = delim.opening.utf8
+    let suffix = delim.closing.utf8
+    guard utf8.prefix(prefix.count).elementsEqual(prefix),
+          utf8.suffix(suffix.count).elementsEqual(suffix) else { return nil }
+
+    return String(utf8.dropFirst(prefix.count).dropLast(suffix.count))
+  }
+  for d in Delimiter.allCases {
+    if let contents = stripDelimiter(d) {
+      return (contents, d)
+    }
+  }
+  fatalError("No valid delimiters")
+}
+
+/// Attempt to lex a regex literal between `start` and `end`, returning either
+/// the contents and pointer from which to resume lexing, or an error.
 public func _lexRegex(
-  _ start: UnsafeRawPointer
-) throws -> (String, SyntaxOptions)? {
+  start: UnsafeRawPointer, end: UnsafeRawPointer
+) -> Result<(contents: String, Delimiter, end: UnsafeRawPointer), LexError> {
```

Comment:
Why a `Result`? I think we can be throwing in everything except the C wrapper shim which can catch the result and pick a common handling strategy.

Reply:
I really only added it so the error could be strongly typed as a `LexError` to make sure that libswift can pick up the resumption point in the case of an error. I'm happy to make it throwing and have libswift catch the LexError specifically though.

Reply:
The mapping logic should remain in the package. This isn't the function directly called by compiler, so it can throw. We can later look at refactoring or sharing logic with later down the pipeline.